# Homework 3: 词性标注与文本分类
学号：16307130194
姓名：陈中钰

In [1]:
%matplotlib inline

导入nltk库和nltk.corpus中的brown语料：

In [2]:
import nltk
# nltk.download()
# from nltk.book import *
from nltk.corpus import brown
print(brown.categories())

['adventure', 'belles_lettres', 'editorial', 'fiction', 'government', 'hobbies', 'humor', 'learned', 'lore', 'mystery', 'news', 'religion', 'reviews', 'romance', 'science_fiction']


## 1. brown语料库处理

从brown语料库中导入使用universal tagset标注的词语数据：

In [3]:
tagged_words = brown.tagged_words(tagset='universal')
tagged_words[0]

('The', 'DET')

数据的数量：

In [4]:
len(tagged_words)

1161192

数据中使用的全部标注tag：

In [5]:
set([tag for (word, tag) in tagged_words])

{'.',
 'ADJ',
 'ADP',
 'ADV',
 'CONJ',
 'DET',
 'NOUN',
 'NUM',
 'PRON',
 'PRT',
 'VERB',
 'X'}

写程序处理布朗语料库，找到以下问题的答案：

### 1.1 哪些名词常以它们复数形式而不是它们的单数形式出现？（只考虑常规的复数形式，-s后缀形式的）。

首先统计所有NOUN的词频：

In [6]:
fdist = nltk.FreqDist([word for (word, tag) in tagged_words if tag == 'NOUN'])

然后要判断一个NOUN是否符合常规的复数形式：词本身不以's'结尾，是单数形式，而且加上's'后的复数形式的词也存在。对这些符合要求的词，计算它的复数形式的出现次数占两种形式的出现次数的比例，按照比例从大到小排序，并显示前100个词。

In [7]:
words = []
for word in fdist:
    if not word.endswith('s') and word + 's' in fdist:
        words.append((word, fdist[word + 's'] / (fdist[word] + fdist[word + 's'])))
words.sort(key=lambda w:w[1], reverse=True)

words[0:100]

[('Corp', 0.9882352941176471),
 ('headquarter', 0.9824561403508771),
 ('Catholic', 0.9705882352941176),
 ('stair', 0.9583333333333334),
 ('relative', 0.9545454545454546),
 ('tear', 0.9411764705882353),
 ('mean', 0.94),
 ('employe', 0.9375),
 ('Motor', 0.9347826086956522),
 ('stockholder', 0.9285714285714286),
 ('Persian', 0.9230769230769231),
 ('investor', 0.9166666666666666),
 ('rib', 0.9166666666666666),
 ('microorganism', 0.9166666666666666),
 ('Height', 0.9166666666666666),
 ('Mill', 0.9130434782608695),
 ('kenning', 0.9090909090909091),
 ('Pop', 0.9090909090909091),
 ('Mar', 0.9047619047619048),
 ('assessor', 0.9047619047619048),
 ('ill', 0.9),
 ('subordinate', 0.9),
 ('Relation', 0.9),
 ('gut', 0.9),
 ('elder', 0.9),
 ('Brook', 0.8888888888888888),
 ('Result', 0.8888888888888888),
 ('resource', 0.8888888888888888),
 ('recruit', 0.8888888888888888),
 ('accommodation', 0.8888888888888888),
 ('Scot', 0.8888888888888888),
 ('survivor', 0.8888888888888888),
 ('bun', 0.8888888888888888

### 1.2 哪个词的不同词性标记数目最多？

使用set()统计每个词的词性标记，并输出词性标记数目最多的词语：

In [8]:
pos = {}
for (word, tag) in tagged_words:
    if word not in pos:
        pos[word] = set(tag)
    else:
        pos[word].add(tag)
max_count = 0
max_count_word = ''
for word in pos:
    if len(pos[word]) > max_count:
        max_count, max_count_word = len(pos[word]), word
max_count, max_count_word

(9, 'damn')

词性标记数目最多的词语是'damn'，拥有9种词性标记。

### 1.3 按频率递减的顺序列出标记。前20个最频繁的词性标记代表什么？

使用FreqDist()统计词性标记的数量，并用most_common()按照词性标记的数量从大到小的顺序输出词性标记：

In [9]:
fdist = nltk.FreqDist([tag for (word, tag) in tagged_words])
fdist.most_common(20)

[('NOUN', 275558),
 ('VERB', 182750),
 ('.', 147565),
 ('ADP', 144766),
 ('DET', 137019),
 ('ADJ', 83721),
 ('ADV', 56239),
 ('PRON', 49334),
 ('CONJ', 38151),
 ('PRT', 29829),
 ('NUM', 14874),
 ('X', 1386)]

其中，'NOUN'表示名词，'VERB'表示动词，'.'表示标点符号，'ADP'表示adpositions(prepositions and postpositions)，'DET'表示限定词，'ADJ'表示形容词，'ADV'表示副词，'PRON'表示代词，'CONJ'表示连词，'PRT'表示particles or other function words，'NUM'表示数词，'X'表示其他词性。

### 1.4 名词后面最常见的是哪些词性标记？这些标记代表什么？

使用bigrams()生成二元词组，过滤出前一个词为名词的二元词组，并用FreqDist()统计这些二元词组的词性标记数量，用most_common()按数量从多到少的顺序输出词性标记：

In [10]:
word_tag_pairs = nltk.bigrams(tagged_words)
noun_post = [post[1] for (pre, post) in word_tag_pairs if pre[1] == 'NOUN']
fdist = nltk.FreqDist(noun_post)
fdist.most_common()

[('.', 78326),
 ('ADP', 67460),
 ('VERB', 43819),
 ('NOUN', 41309),
 ('CONJ', 16451),
 ('ADV', 7307),
 ('PRON', 5492),
 ('PRT', 4940),
 ('DET', 4504),
 ('ADJ', 3603),
 ('NUM', 2255),
 ('X', 92)]

其中，'.'表示标点符号，'ADP'表示adpositions(prepositions and postpositions)，'VERB'表示动词，'NOUN'表示名词，'CONJ'表示连词，'ADV'表示副词，'PRON'表示代词，'PRT'表示particles or other function words，'DET'表示限定词，'ADJ'表示形容词，'NUM'表示数词，'X'表示其他词性。

## 2. nltk分类器

使用nltk提供的任意分类器，以及你能想到的特征，建立一个名字性别分类器。从将名字语料库分成3个子集开始：400个词为测试集，400个词为开发集，剩余的个词为训练集。然后从示例的名字性别分类器开始，逐步改善。使用开发集检查你的进展。一旦你对你的分类器感到满意，在测试集上检查它的最终性能。相比在开发测试集上的性能，它在测试集上的性能如何？

从nltk的corpus中导入names数据，给名字标注上性别的标签，获得总体数据，并shuffle数据：

In [11]:
from nltk.corpus import names
import random

labeled_names = ([(name, 'male') for name in names.words('male.txt')] +
                 [(name, 'female') for name in names.words('female.txt')])
random.seed(1)
random.shuffle(labeled_names)
len(labeled_names)

7944

从总体数据中划分出400个数据作为开发集、400个数据作为测试集，剩下的数据作为训练集：

In [12]:
dev_amount, test_amount = 400, 400

dev_names = labeled_names[:dev_amount]
test_names = labeled_names[dev_amount:dev_amount + test_amount]
train_names = labeled_names[dev_amount + test_amount:]
len(train_names), len(dev_names), len(test_names)

(7144, 400, 400)

### 2.1 以名字最后一个字母为特征

提取最后一个字母为特征：

In [13]:
def get_features(word):
    return {'last1letter': word[-1]}
get_features('Alice')

{'last1letter': 'e'}

使用训练集训练一个简单贝叶斯分类器，并用开发集测试精度：

In [14]:
train_set = [(get_features(name), label) for name, label in train_names]
dev_set = [(get_features(name), label) for name, label in dev_names]
classifier = nltk.NaiveBayesClassifier.train(train_set)
nltk.classify.accuracy(classifier, dev_set)

0.775

从开发集中筛选出预测错误的数据，并进行分析：

In [15]:
errors = []
for name, label in dev_names:
    pred = classifier.classify(get_features(name))
    if pred != label:
        errors.append((name, label, pred))
sorted(errors)

[('Alexis', 'female', 'male'),
 ('Ambrosi', 'male', 'female'),
 ('Anthony', 'male', 'female'),
 ('Archie', 'male', 'female'),
 ('Ardeen', 'female', 'male'),
 ('Babs', 'female', 'male'),
 ('Barney', 'male', 'female'),
 ('Broddy', 'male', 'female'),
 ('Candis', 'female', 'male'),
 ('Casey', 'male', 'female'),
 ('Cherin', 'female', 'male'),
 ('Chevy', 'male', 'female'),
 ('Conway', 'male', 'female'),
 ('Danny', 'male', 'female'),
 ('Dell', 'male', 'female'),
 ('Dennie', 'male', 'female'),
 ('Devan', 'female', 'male'),
 ('Doloritas', 'female', 'male'),
 ('Doralynn', 'female', 'male'),
 ('Dot', 'female', 'male'),
 ('Drake', 'male', 'female'),
 ('Eddy', 'male', 'female'),
 ('Eleanor', 'female', 'male'),
 ('Fan', 'female', 'male'),
 ('Felicdad', 'female', 'male'),
 ('Gene', 'male', 'female'),
 ('Germain', 'female', 'male'),
 ('Giovanni', 'male', 'female'),
 ('Glynis', 'female', 'male'),
 ('Griffith', 'male', 'female'),
 ('Gwenn', 'female', 'male'),
 ('Harvey', 'male', 'female'),
 ('Helmuth', 

仅以最后一个字母来判断性别，精度并不是很高。通过分析错误预测的数据可以发现，以'ly'、'yn'结尾的一般是女性，而以'ie'结尾的一般是男性，因此可以考虑以最后两个字母作为特征。

### 2.2 以名字最后一个、最后两个字母为特征

提取最后一个、最后两个字母为特征：

In [16]:
def get_features(word):
    return {'last1letter': word[-1], 'last2letter': word[-2:]}
get_features('Alice')

{'last1letter': 'e', 'last2letter': 'ce'}

使用训练集训练一个简单贝叶斯分类器，并用开发集测试精度：

In [17]:
train_set = [(get_features(name), label) for name, label in train_names]
dev_set = [(get_features(name), label) for name, label in dev_names]
classifier = nltk.NaiveBayesClassifier.train(train_set)
nltk.classify.accuracy(classifier, dev_set)

0.7925

精度提高了将近2%。接着从开发集中筛选出预测错误的数据，并进行分析：

In [18]:
errors = []
for name, label in dev_names:
    pred = classifier.classify(get_features(name))
    if pred != label:
        errors.append((name, label, pred))
sorted(errors)

[('Aeriel', 'female', 'male'),
 ('Aggy', 'female', 'male'),
 ('Alexis', 'female', 'male'),
 ('Ambrosi', 'male', 'female'),
 ('Angy', 'female', 'male'),
 ('Annabal', 'female', 'male'),
 ('Annabel', 'female', 'male'),
 ('Anthony', 'male', 'female'),
 ('Archie', 'male', 'female'),
 ('Ardeen', 'female', 'male'),
 ('Babs', 'female', 'male'),
 ('Bamby', 'female', 'male'),
 ('Barbey', 'female', 'male'),
 ('Broddy', 'male', 'female'),
 ('Candis', 'female', 'male'),
 ('Cecil', 'female', 'male'),
 ('Ceciley', 'female', 'male'),
 ('Cherin', 'female', 'male'),
 ('Christel', 'female', 'male'),
 ('Clary', 'female', 'male'),
 ('Courtney', 'female', 'male'),
 ('Danny', 'male', 'female'),
 ('Dennie', 'male', 'female'),
 ('Devan', 'female', 'male'),
 ('Doloritas', 'female', 'male'),
 ('Dorothy', 'female', 'male'),
 ('Dot', 'female', 'male'),
 ('Eddy', 'male', 'female'),
 ('Eleanor', 'female', 'male'),
 ('Fan', 'female', 'male'),
 ('Felicdad', 'female', 'male'),
 ('Gene', 'male', 'female'),
 ('Germain', 

通过分析错误预测的数据可以发现，以'Ma'、'Do'、'Ba'开头的都是女性，因此可以考虑以开头两个字母作为特征。

### 2.3 以名字最后一个、最后两个、前两个字母为特征

提取最后一个、最后两个、前两个字母为特征：

In [19]:
def get_features(word):
    return {'last1letter': word[-1], 'last3letter': word[-2:], 'first2letter': word[:2]}
get_features('Alice')

{'last1letter': 'e', 'last3letter': 'ce', 'first2letter': 'Al'}

使用训练集训练一个简单贝叶斯分类器，并用开发集测试精度：

In [20]:
train_set = [(get_features(name), label) for name, label in train_names]
dev_set = [(get_features(name), label) for name, label in dev_names]
classifier = nltk.NaiveBayesClassifier.train(train_set)
nltk.classify.accuracy(classifier, dev_set)

0.825

从开发集中筛选出预测错误的数据：

In [21]:
errors = []
for name, label in dev_names:
    pred = classifier.classify(get_features(name))
    if pred != label:
        errors.append((name, label, pred))
sorted(errors)

[('Alexis', 'female', 'male'),
 ('Ambrosi', 'male', 'female'),
 ('Anthony', 'male', 'female'),
 ('Archie', 'male', 'female'),
 ('Ardeen', 'female', 'male'),
 ('Babs', 'female', 'male'),
 ('Bamby', 'female', 'male'),
 ('Barbey', 'female', 'male'),
 ('Broddy', 'male', 'female'),
 ('Candis', 'female', 'male'),
 ('Casey', 'male', 'female'),
 ('Cherin', 'female', 'male'),
 ('Christel', 'female', 'male'),
 ('Clary', 'female', 'male'),
 ('Conway', 'male', 'female'),
 ('Danny', 'male', 'female'),
 ('Dennie', 'male', 'female'),
 ('Devan', 'female', 'male'),
 ('Doloritas', 'female', 'male'),
 ('Dot', 'female', 'male'),
 ('Eddy', 'male', 'female'),
 ('Eleanor', 'female', 'male'),
 ('Felicdad', 'female', 'male'),
 ('Gene', 'male', 'female'),
 ('Germain', 'female', 'male'),
 ('Giovanni', 'male', 'female'),
 ('Haleigh', 'female', 'male'),
 ('Havivah', 'female', 'male'),
 ('Hewie', 'male', 'female'),
 ('Ingeborg', 'female', 'male'),
 ('Jean', 'female', 'male'),
 ('Julie', 'male', 'female'),
 ('Keith'

精度提高了超过3%，达到了82.5%，效果还是很不错的。在测试集上检查模型的最终性能：

In [22]:
test_set = [(get_features(name), label) for name, label in test_names]
nltk.classify.accuracy(classifier, test_set)

0.8175

模型在开发测试集上的性能为82.5%，而它在测试集上的性能为81.75%，要稍微低一点。因为使用了训练集来训练模型，并通过在开发集上测试得到的精度来选择一个最好的模型，以防止模型过拟合。而测试集代表的是模型从来没有见过的、更接近现实的数据，所以精度会稍微低一点。